### ISSUE TITLE
[REGRESSION]: electron tests sometimes freezes for 30 seconds

### ISSUE THREAD

**Issue by ZFail** [microsoft/playwright#27523]
**Context:**
- GOOD Playwright Version: 1.36.2
- BAD Playwright Version: 1.38.1
- Operating System: Windows 10

**Code Snippet**

tested on main repository with command

```
npx playwright test electron-app.spec.ts
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.

Running 22 tests using 1 worker

  ok 1 tests\electron\electron-app.spec.ts:22:5 › should fire close event (1.2s)
  ok 2 tests\electron\electron-app.spec.ts:34:5 › should dispatch ready event (137ms)
  ok 3 tests\electron\electron-app.spec.ts:46:5 › should script application (145ms)
  ok 4 tests\electron\electron-app.spec.ts:51:5 › should preserve args (137ms)
  ok 5 tests\electron\electron-app.spec.ts:58:5 › should return windows (349ms)
  ok 6 tests\electron\electron-app.spec.ts:63:5 › should evaluate handle (147ms)
  ok 7 tests\electron\electron-app.spec.ts:68:5 › should route network (380ms)
  ok 8 tests\electron\electron-app.spec.ts:81:5 › should support init script (355ms)
  ok 9 tests\electron\electron-app.spec.ts:88:5 › should expose function (362ms)
  ok 10 tests\electron\electron-app.spec.ts:95:5 › should wait for first window (351ms)
  ok 11 tests\electron\electron-app.spec.ts:104:5 › should have a clipboard instance (151ms)
  ok 12 tests\electron\electron-app.spec.ts:111:5 › should test app that opens window fast (136ms)
  ok 13 tests\electron\electron-app.spec.ts:115:5 › should return browser window (378ms)
  ok 14 tests\electron\electron-app.spec.ts:122:5 › should bypass csp (453ms)
  ok 15 tests\electron\electron-app.spec.ts:137:5 › should create page for browser view (524ms)
  ok 16 tests\electron\electron-app.spec.ts:150:5 › should return same browser window for browser view pages (575ms)
  ok 17 tests\electron\electron-app.spec.ts:171:5 › should record video (1.5s)
  ok 18 tests\electron\electron-app.spec.ts:183:5 › should be able to get the first window when with a delayed navigation (2.3s)
  ok 19 tests\electron\electron-app.spec.ts:192:5 › should detach debugger on app-initiated exit (158ms)
  ok 20 tests\electron\electron-app.spec.ts:201:5 › should run pre-ready apis (140ms)
  ok 21 tests\electron\electron-app.spec.ts:205:5 › should resolve app path for folder apps (136ms)
  ok 22 tests\electron\electron-app.spec.ts:211:5 › should return app name / version from manifest (134ms)

  22 passed (41.1s)

```

**Describe the bug**

In the log you can see that the tests pass on average in 10 seconds, what happens in the remaining 30 seconds is unclear. Playwright freezes for 30 seconds after the last test is completed.

sometimes this command ends in 10 seconds, but rarely
previous version of playwright also ends this test in 10 seconds

**Comment by ZFail** [comment]
I ran the tests on another computer with Windows 10 and got the same results
repro:
```
git clone https://github.com/microsoft/playwright.git
cd playwright
git checkout tags/v1.38.1 -b electron-hang-reproducer
npm ci
npm run build
npx playwright test electron-app.spec.ts

Running 22 tests using 1 worker

  ✓  1 tests\electron\electron-app.spec.ts:22:5 › should fire close event (1.2s)
  ✓  2 tests\electron\electron-app.spec.ts:34:5 › should dispatch ready event (129ms)
  ✓  3 tests\electron\electron-app.spec.ts:46:5 › should script application (138ms)
  ✓  4 tests\electron\electron-app.spec.ts:51:5 › should preserve args (132ms)
  ✓  5 tests\electron\electron-app.spec.ts:58:5 › should return windows (359ms)
  ✓  6 tests\electron\electron-app.spec.ts:63:5 › should evaluate handle (136ms)
  ✓  7 tests\electron\electron-app.spec.ts:68:5 › should route network (367ms)
  ✓  8 tests\electron\electron-app.spec.ts:81:5 › should support init script (392ms)
  ✓  9 tests\electron\electron-app.spec.ts:88:5 › should expose function (355ms)
  ✓  10 tests\electron\electron-app.spec.ts:95:5 › should wait for first window (362ms)
  ✓  11 tests\electron\electron-app.spec.ts:104:5 › should have a clipboard instance (137ms)
  ✓  12 tests\electron\electron-app.spec.ts:111:5 › should test app that opens window fast (132ms)
  ✓  13 tests\electron\electron-app.spec.ts:115:5 › should return browser window (378ms)
  ✓  14 tests\electron\electron-app.spec.ts:122:5 › should bypass csp (417ms)
  ✓  15 tests\electron\electron-app.spec.ts:137:5 › should create page for browser view (500ms)
  ✓  16 tests\electron\electron-app.spec.ts:150:5 › should return same browser window for browser view pages (506ms)
  ✓  17 tests\electron\electron-app.spec.ts:171:5 › should record video (1.4s)
  ✓  18 tests\electron\electron-app.spec.ts:183:5 › should be able to get the first window when with a delayed navigation (2.3s)
  ✓  19 tests\electron\electron-app.spec.ts:192:5 › should detach debugger on app-initiated exit (139ms)
  ✓  20 tests\electron\electron-app.spec.ts:201:5 › should run pre-ready apis (122ms)
  ✓  21 tests\electron\electron-app.spec.ts:205:5 › should resolve app path for folder apps (125ms)
  ✓  22 tests\electron\electron-app.spec.ts:211:5 › should return app name / version from manifest (124ms)

  22 passed (40.7s)
```

with old version 1.36.2 all is ok:
```
git clone https://github.com/microsoft/playwright.git
cd playwright
git checkout tags/v1.36.2 -b electron-hang-reproducer2
npm ci
npm run build
npx playwright test electron-app.spec.ts

Running 22 tests using 1 worker

  ✓  1 tests\electron\electron-app.spec.ts:22:5 › should fire close event (1.4s)
  ✓  2 tests\electron\electron-app.spec.ts:34:5 › should dispatch ready event (128ms)
  ✓  3 tests\electron\electron-app.spec.ts:46:5 › should script application (129ms)
  ✓  4 tests\electron\electron-app.spec.ts:51:5 › should preserve args (129ms)
  ✓  5 tests\electron\electron-app.spec.ts:58:5 › should return windows (369ms)
  ✓  6 tests\electron\electron-app.spec.ts:63:5 › should evaluate handle (134ms)
  ✓  7 tests\electron\electron-app.spec.ts:68:5 › should route network (373ms)
  ✓  8 tests\electron\electron-app.spec.ts:81:5 › should support init script (366ms)
  ✓  9 tests\electron\electron-app.spec.ts:88:5 › should expose function (366ms)
  ✓  10 tests\electron\electron-app.spec.ts:95:5 › should wait for first window (349ms)
  ✓  11 tests\electron\electron-app.spec.ts:104:5 › should have a clipboard instance (133ms)
  ✓  12 tests\electron\electron-app.spec.ts:111:5 › should test app that opens window fast (130ms)
  ✓  13 tests\electron\electron-app.spec.ts:115:5 › should return browser window (344ms)
  ✓  14 tests\electron\electron-app.spec.ts:122:5 › should bypass csp (407ms)
  ✓  15 tests\electron\electron-app.spec.ts:137:5 › should create page for browser view (500ms)
  ✓  16 tests\electron\electron-app.spec.ts:150:5 › should return same browser window for browser view pages (500ms)
  ✓  17 tests\electron\electron-app.spec.ts:171:5 › should record video (1.3s)
  ✓  18 tests\electron\electron-app.spec.ts:183:5 › should be able to get the first window when with a delayed navigation (2.2s)
  ✓  19 tests\electron\electron-app.spec.ts:192:5 › should detach debugger on app-initiated exit (129ms)
  ✓  20 tests\electron\electron-app.spec.ts:201:5 › should run pre-ready apis (123ms)
  ✓  21 tests\electron\electron-app.spec.ts:205:5 › should resolve app path for folder apps (125ms)
  ✓  22 tests\electron\electron-app.spec.ts:211:5 › should return app name / version from manifest (123ms)

  22 passed (10.2s)
```

**Comment by dgozman** [comment]
@ZFail If you are running Playwright's own tests, you should do it like this:

```ts
npm run etest electron-app.spec
```

That said, what's your actual problem? Are you running Playwright's own tests for some reason? Or perhaps you do have some real electron tests that are slow? Note that it also depends on the Electron itself, so a repro would be very apprciated.

**Comment by ZFail** [comment]
> npm run etest electron-app.spec

This command fixes strange freeze. Now test passes in 10 seconds.

I found line in config that causes this freeze, if i comment out [this line](https://github.com/microsoft/playwright/blob/47434209996cfd19d7c913caa7ddf8e1377e1d99/tests/electron/playwright.config.ts#L56C7-L56C7) (coverageName: 'electron'), then tests freezes and the end for 30 seconds.

I don’t understand what this line does and what should I do to fix the tests in my project

> Are you running Playwright's own tests for some reason? 

Cause its simple reproducer of my problem.

>Or perhaps you do have some real electron tests that are slow? 

I encountered problem in tests when updating the version of playwright in my project, then I tried to run tests in the playwright repository and realized that the problem also exists in this repo.

**Comment by dgozman** [comment]
> This command fixes strange freeze. Now test passes in 10 seconds.

That's great!

> I don’t understand what this line does and what should I do to fix the tests in my project

This line is completely unrelated to your project, it just collects some coverage for Playwright APIs from our internal tests. I don't think it's worth the time figuring out why running the wrong command makes things freeze in our repo.

> I encountered problem in tests when updating the version of playwright in my project, then I tried to run tests in the playwright repository and realized that the problem also exists in this repo.

Thank you for the explanation! I think it would still be beneficial to get an actual repro, because we don't observe this issue in Playwright's own tests. Most likely, electron fails to exit for you. You can try running tests with `DEBUG=pw:browser*` environment variable - perhaps that would show some logs from electron that explain the issue?

**Comment by ZFail** [comment]
create repro https://github.com/ZFail/electron-playwright-repro.git
```
npm run test

> electron-quick-start@1.0.0 test
> playwright test


Running 1 test using 1 worker

  ✓  1 [electron] › tests.spec.ts:3:5 › check initial state (412ms)

  1 passed (30.8s)
```

**Comment by dgozman** [comment]
I can repro this on Windows, thank you.

**Comment by planetsLightningArrester** [comment]
I'm having the same issue: the Electron test passes but hangs for 30s before finishing. I'm unsure if it's the same root cause as OP.

What causes the hanging in my case is the call of `await electronApp.close()`. Removing it the test passes as expected.

This issue arose after upgrading from v1.35.1 to v1.40.1. I wonder if some process is running in Electron that `.close()` won't kill.

**Comment by mxschmitt** [comment]
Looks like this is fixed in recent Playwright versions like 1.42. I was able to bisect it to:

https://github.com/microsoft/playwright/compare/0c4de1f4ce869af86c99444550c1fcc207fc6b3a...817a130cdc71140acfafb3aafbab7f3e267a9f08

1.40.0-alpha-nov-2-2023 Good
1.40.0-alpha-nov-1-2023 Bad

@planetsLightningArrester could you file a new issue with a reproduction? With the repro from @ZFail I'm not able to reproduce it after 1.40.

**Comment by planetsLightningArrester** [comment]
Hi @mxschmitt

I can still reproduce it with the version latest version of Electron and Playwright.
Electron: 28.2.1
Playwright: 1.41.2

I created a repo here to reproduce the issue: https://github.com/planetsLightningArrester/electron-hang-demo

## Reproducing
```bash
git clone git@github.com:planetsLightningArrester/electron-hang-demo.git
npm ci
npm test
```

This will pass in under 1s. But if we uncomment the line below in the file `src/__tests__/test.spec.ts`

```typescript
// Un-commenting the code below makes the test hang for 30s
test.afterAll(async () => {
  await electronApp.close()
})
```

The test passes **only after hanging 30s**. I tried to comment most of `src/main.ts`, but I got the same results. So I left everything there.

I didn't create a new issue because I really think this is the same. But you know better and if you disagree I'll file a new issue :)

**Comment by mxschmitt** [comment]
Hi @planetsLightningArrester, I tried to reproduce it, but for me its all passing, I tried with and without `electronApp.close()` see here: 


https://github.com/microsoft/playwright/assets/17984549/2ece56d8-05a2-4c23-9b66-d8f8b1875723


Maybe try also `npx playwright --version` or can you set the `$Env:DEBUG="pw:api,pw:browser,pw:protocol"` env var before? This might give us some more debug output.

**Comment by planetsLightningArrester** [comment]
Hi @mxschmitt

Thanks for testing it, and I'm sorry for the late response. Indeed it works on Windows but fails on WSL. Can you please check if you can reproduce it in WSL? I'm using Windows 11H1 and WSL with Ubuntu 22.04.3

Since this looks OS-related, I can open a new issue if you think it's necessary

**Comment by mxschmitt** [comment]
I can repro on WSL.

bad   1.37.0-alpha-jul-25-2023 
good 1.37.0-alpha-jul-24-2023 
Bisect: https://github.com/microsoft/playwright/compare/207086180781e1daee52bcba0aefe3bed1b87b2e...5c41c6e79923d9cb0953fd51d4a37ee0f20e06af
